hack/generator/pkg/codegen/pipeline_convert_allof_and_oneof_to_objects.go

```diff
@@ -118,63 +122,72 @@ type synthesizer struct {
 	defs         astmodel.Types
 }
 
-func (s synthesizer) oneOfObject(oneOf astmodel.OneOfType) (astmodel.Type, error) {
-	// If there's more than one option, synthesize a type.
-	// Note that this is required because Kubernetes CRDs do not support OneOf the same way
-	// OpenAPI does, see https://github.com/Azure/k8s-infra/issues/71
-	var properties []*astmodel.PropertyDefinition
-
-	propertyDescription := "Mutually exclusive with all other properties"
-	err := oneOf.Types().ForEachError(func(t astmodel.Type, i int) error {
-		prop, err := s.convertToOneOfProperty(i, t)
-		if err != nil {
-			return err
-		}
+type propertyNames struct {
+	golang astmodel.PropertyName
+	json   string
 
-		prop = prop.MakeOptional()
-		prop = prop.WithDescription(propertyDescription)
+	// used to resolve conflicts:
+	isGoodName bool
+	depth      int
+}
 
-		properties = append(properties, prop)
-		return nil
-	})
+func (ns propertyNames) betterThan(other propertyNames) bool {
+	if ns.isGoodName && !other.isGoodName {
+		return true
+	}
 
-	if err != nil {
-		return nil, err
+	if !ns.isGoodName && other.isGoodName {
+		return false
 	}
 
-	objectType := astmodel.NewObjectType().WithProperties(properties...)
-	objectType = objectType.WithFunction(astmodel.NewOneOfJSONMarshalFunction(objectType, s.idFactory))
+	// both are good or !good
+	// return the name closer to the top
+	// (i.e. “lower” in inheritance hierarchy)
```

Comment:
minor: Add a comment explaining what this does?

I _think_ it's basically saying that the property closest to the "leaf" of the oneof/allof graph is the one whose name wins? Why is that a desirable characteristic?

Reply:
It's the other way around, “higher” in the inheritance hierarchy = deeper in the graph. I’ll expand this a bit more